### PR TITLE
perf: take ordermap instead of std hashmap

### PR DESCRIPTION
**Pull request by lishihai9017**
@zhangjinpeng1987 @BusyJay @siddontang @overvenus , PTAL

### CODE REVIEW

File: src/util/codec/table.rs [hunk 1]
```diff
@@ -325,175 +324,200 @@ pub fn cut_idx_key(key: Vec<u8>, col_ids: &[i64]) -> Result<(RowColsDict, Option
     Ok((RowColsDict::new(meta_map, key), handle))
 }
 
-#[cfg(test)]
```
Comment (siddontang) [inline src/util/codec/table.rs]: why remove the test?

**Comment by siddontang** [comment]
any performance increased?

File: src/util/config.rs [hunk 2]
```diff
@@ -468,7 +468,7 @@ mod test {
         }
 
         let map = parse_store_labels("").unwrap();
-        assert!(map.is_empty());
+        assert_eq!(map.is_empty());
```
Comment (overvenus) [inline src/util/config.rs]: Should be `assert!()`

File: src/util/xeval/evaluator.rs [hunk 3]
```diff
@@ -297,9 +297,10 @@ impl Evaluator {
 
     fn decode_value_list(&mut self, value_list_expr: &Expr) -> Result<&Vec<Datum>> {
         let p = value_list_expr as *const Expr as isize;
-        let decoded = try!(self.cached_value_list
+        let v = try!(value_list_expr.get_val().decode());
+        let decoded = self.cached_value_list
             .entry(p)
-            .or_try_insert_with(|| value_list_expr.get_val().decode()));
+            .or_insert(v);
```
Comment (overvenus) [inline src/util/xeval/evaluator.rs]: Although it works, could you lazy decode `value_list_expr.get_val()`?

File: Cargo.toml [hunk 4]
```diff
@@ -59,6 +59,7 @@ getopts = "0.2"
 regex = "0.1"
 fnv = "1.0"
 sys-info = "0.4.1"
+ordermap = "0.2.9"
```
Comment (BusyJay) [inline Cargo.toml]: Use "0.2" instead.

File: src/server/coprocessor/endpoint.rs [hunk 5]
```diff
@@ -673,16 +673,19 @@ impl SelectContextCore {
                 for cond_col in cond_col_map.keys() {
                     aggr_cols_map.remove(cond_col);
                 }
-                aggr_cols = aggr_cols_map.drain().map(|(_, v)| v).collect();
+                aggr_cols = aggr_cols_map.clone().into_iter().map(|(_, v)| v).collect();
```
Comment (BusyJay) [inline src/server/coprocessor/endpoint.rs]: Doesn't have to clone.

File: src/util/collections.rs [hunk 6]
```diff
@@ -0,0 +1,22 @@
+// Copyright 2016 PingCAP, Inc.
```
Comment (BusyJay) [inline src/util/collections.rs]: s/2016/2017/

File: src/util/collections.rs [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub use fnv::FnvHashSet as HashSet;
+pub use fnv::FnvBuildHasher as BuildHasherDefault;
+// pub use fnv::FnvHashMap as HashMap;
+// pub use std::collections::hash_map::Entry as HashMapEntry;
+// pub use std::collections::hash_map::Values as HashMapValues;
+
+pub use ordermap::OrderMap as HashMap;
```
Comment (BusyJay) [inline src/util/collections.rs]: I don't think it's appropriate to call it `HashMap`.

File: src/util/collections.rs [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub use fnv::FnvHashSet as HashSet;
+pub use fnv::FnvBuildHasher as BuildHasherDefault;
+// pub use fnv::FnvHashMap as HashMap;
```
Comment (BusyJay) [inline src/util/collections.rs]: Just remove the code.

File: src/util/codec/table.rs [hunk 9]
```diff
@@ -264,7 +265,7 @@ impl RowColsDict {
     }
 
     pub fn is_empty(&self) -> bool {
-        self.cols.is_empty()
+        self.cols.len() == 0
```
Comment (andelf) [inline src/util/codec/table.rs]: is_empty()

File: src/util/collections.rs (hunk 7 above)
Comment (andelf) [inline src/util/collections.rs]: We've introduced too many `as` renames. Can't we use path expr?

ordermap::Entry instead of `as MapEntry`

Comment (siddontang) [inline src/util/collections.rs]: there may be lots of changes if not using HashMap. We can do so first, then rename a proper name later. Maybe we can use Map directly.

Comment (BusyJay) [inline src/util/collections.rs]: It's ok to rename it in this pr.

File: tests/raft/test_raft.rs [hunk 10]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2016 PingCAP, Inc.
+// Copyright 2017 PingCAP, Inc.
```
Comment (siddontang) [inline tests/raft/test_raft.rs]: no need to change the existing file copyright.

File: src/server/coprocessor/endpoint.rs [hunk 11]
```diff
@@ -673,16 +673,17 @@ impl SelectContextCore {
...
-            cond_cols = cond_col_map.drain().map(|(_, v)| v).collect();
+            cond_cols = cond_col_map.into_iter().map(|(_, v)| v).collect();
 
             // get topn cols
             let mut topn_col_map = HashMap::default();
             for item in sel.get_order_by() {
                 try!(collect_col_in_expr(&mut topn_col_map, select_cols, item.get_expr()))
             }
-            topn_cols = topn_col_map.drain().map(|(_, v)| v).collect();
+            topn_cols = topn_col_map.clone().into_iter().map(|(_, v)| v).collect();
```
Comment (siddontang) [inline src/server/coprocessor/endpoint.rs]: why clone here?

File: src/util/collections.rs [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub use fnv::FnvHashSet as HashSet;
+pub use fnv::FnvBuildHasher as BuildHasherDefault;
+pub use ordermap::OrderMap as HashMap;
```
Comment (BusyJay) [inline src/util/collections.rs]: Just name it `Map` instead.

File: src/util/codec/table.rs [hunk 13]
```diff
@@ -418,35 +416,53 @@ mod test {
...
+        assert!(col_encoded.iter()
+            .all(|(key, value)| datums.get(key).map_or(false, |v| *value == *v)));
 
         let bs = encode_row(vec![], &[]).unwrap();
-        assert!(!bs.is_empty());
+        assert_ne!(bs.len(), 0);
         assert!(bs.as_slice().decode_row(&Default::default(), &cols).unwrap().is_empty());
         datums = cut_row_as_owned(&bs, &col_id_set);
-        assert!(datums.is_empty());
+        assert_eq!(datums.len(), 0);
```
Comment (BusyJay) [inline src/util/codec/table.rs]: Use `is_empty` instead.

File: src/util/collections.rs [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub use fnv::FnvHashSet as HashSet;
```
Comment (siddontang) [inline src/util/collections.rs]: No order set now?

Comment (lishihai9017) [inline src/util/collections.rs]: ordermap crate doesn't provide an OrderSet struct.

File: src/util/codec/table.rs [hunk 15]
```diff
@@ -418,32 +416,50 @@ mod test {
         assert!(!bs.is_empty());
 
         let r = bs.as_slice().decode_row(&Default::default(), &cols).unwrap();
-        assert_eq!(row, r);
+        assert_eq!(row.len(), r.len());
```
Comment (overvenus) [inline src/util/codec/table.rs]: `ordermap` impls `PartialEq`, I think `assert_eq!(row, r)` is enough.

File: src/util/codec/table.rs [hunk 16]
```diff
@@ -418,32 +416,50 @@ mod test {
...
         datums = cut_row_as_owned(&bs, &col_id_set);
         col_encoded.remove(&3);
-        assert_eq!(col_encoded, datums);
+        assert_eq!(col_encoded.len(), datums.len());
+        assert!(col_encoded.iter()
+            .all(|(key, value)| datums.get(key).map_or(false, |v| *value == *v)));
 
         let bs = encode_row(vec![], &[]).unwrap();
-        assert!(!bs.is_empty());
+        assert_ne!(bs.len(), 0);
```
Comment (overvenus) [inline src/util/codec/table.rs]: How about `is_empty()`?

File: src/util/codec/table.rs (hunk 15 above)
Comment (BusyJay) [inline src/util/codec/table.rs]: Why not keep the `assert_eq`?

File: src/util/codec/table.rs (hunk 16 above)
Comment (BusyJay) [inline src/util/codec/table.rs]: Why not use `is_empty`?

File: tests/raft/test_raft.rs [hunk 17]
```diff
@@ -180,8 +181,7 @@ impl Interface {
     fn initial(&mut self, id: u64, ids: &[u64]) {
         if self.raft.is_some() {
             self.id = id;
-            self.prs = RaftHashMap::with_capacity_and_hasher(ids.len(),
-                                                             BuildHasherDefault::default());
+            self.prs = RaftHashMap::with_capacity_and_hasher(ids.len(), RandomState::new());
```
Comment (BusyJay) [inline tests/raft/test_raft.rs]: I think `with_capacity` is enough.

File: src/util/xeval/evaluator.rs [hunk 18]
```diff
@@ -297,9 +297,9 @@ impl Evaluator {
 
     fn decode_value_list(&mut self, value_list_expr: &Expr) -> Result<&Vec<Datum>> {
         let p = value_list_expr as *const Expr as isize;
-        let decoded = try!(self.cached_value_list
+        let decoded = self.cached_value_list
             .entry(p)
-            .or_try_insert_with(|| value_list_expr.get_val().decode()));
+            .or_insert(value_list_expr.get_val().decode().unwrap());
```
Comment (BusyJay) [inline src/util/xeval/evaluator.rs]: Use `match` instead.

Comment (overvenus) [inline src/util/xeval/evaluator.rs]: Try this :wink:  
```rust
        use util::collections::HashMapEntry;

        let decoded = match self.cached_value_list
            .entry(p) {
            HashMapEntry::Occupied(entry) => entry.into_mut(),
            HashMapEntry::Vacant(entry) => {
                let default = try!(value_list_expr.get_val().decode());
                entry.insert(default)
            }
        };
```

File: src/util/collections.rs [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+pub use fnv::FnvBuildHasher as BuildHasherDefault;
+pub use ordermap::OrderMap as HashMap;
+pub use ordermap::Entry as HashMapEntry;
+pub use ordermap::Values as HashMapValues;
```
Comment (zhangjinpeng1987) [inline src/util/collections.rs]: EOF

File: src/util/xeval/evaluator.rs (hunk 18 above)
Comment (lishihai9017) [inline src/util/xeval/evaluator.rs]: Nice

**Comment by zhangjinpeng1987** [comment]
LGTM  please add bench result for core function.

**Comment by ngaut** [comment]
LGTM, Please fix CI.

**Comment by lishihai9017** [comment]
[Bench Map Result](https://github.com/pingcap/tikv/issues/1676#issuecomment-288611580)

```
test collections::bench_fnv_map_large   ... bench:     177,892 ns/iter (+/- 32,867)
test collections::bench_fnv_map_little  ... bench:       1,468 ns/iter (+/- 56)
test collections::bench_ordermap_large  ... bench:     109,889 ns/iter (+/- 4,394)
test collections::bench_ordermap_little ... bench:         986 ns/iter (+/- 36)
```

**Pull request merged**